src/ui/src/mainwindow.cpp

```diff
@@ -1,44 +1,101 @@
 	render_widget_->showMaximized();
 	// shape widget
 	shape_list_widget_ = new ShapeListWidget(this);
 	addDockWidget(Qt::LeftDockWidgetArea, shape_list_widget_);
 	shape_list_widget_->setWindowTitle("Shapes");
 	// metadata widget
-	metadata_widget_ = new MetadataWidget(this);
-	addDockWidget(Qt::LeftDockWidgetArea, metadata_widget_);
-	metadata_widget_->setWindowTitle("Metadata");
+	metadata_widget_           = new MetadataWidget(this);
+	QDockWidget* metadata_dock = new QDockWidget(tr("Metadata"), this);
+	metadata_dock->setWidget(metadata_widget_);
+	addDockWidget(Qt::LeftDockWidgetArea, metadata_dock);
+	connect(this, &MainWindow::ShowMetadata, metadata_widget_,
+	        &MetadataWidget::OnShowMetadata);
+
 	// options widget
 	visualization_options_ = new VisualizationOptionsWidget(this);
 	addDockWidget(Qt::LeftDockWidgetArea, visualization_options_);
 	visualization_options_->setWindowTitle("Options");
 	// signals and slots connection
 	connect(ui_->load_file_btn, &QAction::triggered, this,
 	        &MainWindow::OnLoadFileBtnPressed);
+	connect(visualization_options_,
+	        &VisualizationOptionsWidget::VisualizationOptionsChanged,
+	        render_widget_, &RendererWidget::onVisualizationOptionsChanged);
 }
 
 MainWindow::~MainWindow() { delete ui_; }
 
-void MainWindow::LoadFile(const std::string& file_name) {
-	core().LoadGeometry(file_name);
-	render_widget_->RenderShapes(core().GetShapes());
+void MainWindow::LoadFile(const std::string& trm_file,
+                          const std::string& t2d_file) {
+	core().LoadGeometry(trm_file, t2d_file);
+	const auto loaded_shapes = core().GetFirstFile().GetShapes();
+	const auto loaded_heatmaps = core().GetFirstFile().heatmaps();
+	// Do we really need this assert?
+	// assert(!loaded_shapes.empty() && "no shapes received");
+	if (!loaded_shapes.empty()) {
+		LOG_INFO("Got {} files to render", loaded_shapes.size());
+		// FIXME: Adding shapes shouldn't need to make API context current
+		render_widget_->makeCurrent();
+		scene_->Clear();
+		scene_->AddShapes(loaded_shapes);
+		scene_->AddHeatmaps(loaded_heatmaps);
+		render_widget_->doneCurrent();
+	} else {
+		LOG_WARN("No shaped received from file {}", trm_file);
+	}
 }
 
 void MainWindow::OnLoadFileBtnPressed() {
-	const QString file_name = QFileDialog::getOpenFileName(
-			this, tr("Open File"), QDir::currentPath(), tr("geom (*.cpp)"));
+	const QString trm_file = QFileDialog::getOpenFileName(
+			this, tr("Open trm File"), QDir::currentPath(),
+			tr("geom (*.txt *.TRM);; ALL (*.*)"));
+	const QString t2d_file = QFileDialog::getOpenFileName(
+			this, tr("Open T2D File"), QDir::currentPath(),
+			tr("geom (*.txt *.T2D);; ALL (*.*)"));
 	// TODO: check if file_name is empty (on cancel)
-	LoadFile(file_name.toStdString());
+	if (trm_file.length() && t2d_file.length()) {
+		try {
+			LoadFile(trm_file.toStdString(), t2d_file.toStdString());
+		} catch (...) {
+			QMessageBox messageBox;
+			messageBox.critical(
```

Comment:
`critical` is static member function. Don't call it from intance. So it's pointless to `SetFixedSize` from that object.